web/flat-web/src/utils/CoursewarePreloader/utils.ts

```diff
@@ -0,0 +1,40 @@
+let cachePromise: Promise<Cache> | undefined;
+export function openCache(cacheName: string): Promise<Cache> {
+    return (cachePromise ||= caches.open(cacheName));
+}
+
+class CacheMember {
+    private _flat: Promise<Cache> | undefined;
+
+    public flat(): Promise<Cache> {
+        return (this._flat ||= caches.open("flat"));
+    }
+}
+
+export const cacheMember = new CacheMember();
+
+const contentTypeByExt: Record<string, string> = {
+    ".html": "text/html",
+    ".css": "text/css",
+    ".js": "application/javascript",
+    ".txt": "text/plain",
+    ".json": "application/json",
+    ".xml": "application/xml",
+    ".jpg": "image/jpeg",
+    ".jpeg": "image/jpeg",
+    ".png": "image/png",
+    ".gif": "image/gif",
+    ".svg": "image/svg+xml",
+    ".mp3": "audio/mpeg",
+    ".mp4": "video/mp4",
+    ".woff": "font/woff",
+    ".woff2": "font/woff2",
+    ".ttf": "font/ttf",
+};
+
+// it is guaranteed that the url does not follows '#' or '?'
+export function contentType(url: string): string {
+    const filename = url.substring(url.lastIndexOf("/"));
+    const ext = filename.substring(filename.lastIndexOf("."));
+    return contentTypeByExt[ext] || "text/plain";
```

Comment:
When it is not in the supported range, we give the default value of x. Is it within expectations? Will there be any problem?

If it is not here, can we not cache it?

Reply:
```
let cachePromise: Promise<Cache> | undefined;
export function openCache(cacheName: string): Promise<Cache> {
    return (cachePromise ||= caches.open(cacheName));
}
```

is not equal to

```
class CacheMember {
    private _pptConvert: Cache;
    constructor() {}

    public pptConvert(): Promsie<Cache> {
        return (this._pptConvert ||= caches.open("ppt-convert"));
    }
}
```

If you construct a class object, you need to hold this class instance as a constant and export it everywhere, which is no advantage compared to the original function export.

Reply:
He can help us manage the cache without worrying about intermediate variables being contaminated

Previous writing style:

```ts
let cachePromise: Promise<Cache> | undefined;
export function openCache(cacheName: string): Promise<Cache> {
    return (cachePromise ||= caches.open(cacheName));
}

const flat = openCache("flat");
const web = openCache("web");

flat === web // result => true
```
This is very confusing

What I mean is:

```ts
class CacheMember {
    private _pptConvert: Cache;
    constructor() {}

    public pptConvert(): Promsie<Cache> {
        return (this._pptConvert ||= caches.open("ppt-convert"));
    }
}

export const cacheMember = new CacheMember();
```

The previous practice would also be imported anywhere, and still is. This is not a convincing reason

Reply:
you're right. 

but if you also need export a const, the class is not a good practice in JavaScript. it always confuses beginner `which this mean`.

```
let cachePromise: Promise<Cache> | undefined;
function openCache(cacheName: string): Promise<Cache> {
    return (cachePromise ||= caches.open(cacheName));
}
const flatCache = openCache("flat")
export flatCache;
```

or 

```
class CacheMember {
    private _pptConvert: Cache;
    constructor(private name: string) {}

    public pptConvert(): Promsie<Cache> {
        return (this._pptConvert ||= caches.open(name));
    }
}

export const cacheMember = new CacheMember("flat");
```

Class is not a good way to practice here.

@crimx  recommend https://github.com/alexreardon/memoize-one.

Reply:
Thank you for your suggestion, I made optimization on top of your base

```ts
class CacheBuilder {
    private cache: Promise<Cache> | undefined;

    constructor(private cacheName: string) {}

    public build(): Promsie<Cache> {
        return (this.cache ||= caches.open(this.cacheName));
    }
}

export const cachePPTConvert = new CacheBuilder("ppt-convert").builer;
```

Reply:
Although memoize-one can be used, I don't recommend it. If used here, it will cause the code here to lose its original meaning.

Before: lazy loading
Now: listen for value changes, and if the value changes, I will re-execute